File: ios/PotillusKit/Sources/PotillusKit/Domain/ReportChart.swift
// vim: set et ts=4:
// =============================================================================
// Libellus Potionis - Privacy-Friendly Alcohol Tracker
// Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
// =============================================================================
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program.  If not, see <https://www.gnu.org/licenses/>.
//
// In addition, as permitted by section 7 of the GNU General Public License,
// this program may carry additional permissions; any such permissions that
// apply to it are stated in the accompanying COPYING.md file.
//
// =============================================================================

import Foundation

// =============================================================================
// ReportChart – the report's presentation arithmetic
// =============================================================================
//
// Bar heights, axis-label thinning, donut geometry, category colours. Everything
// the PDF needs in order to LOOK the same on both platforms, and nothing that
// needs a language: no strings, no locale, no formatting of numbers for a reader.
//
// Pinned by test-vectors/report-chart.json, which both suites read.
// =============================================================================

public enum ReportChart {

    /// The smallest bar a non-zero value may draw, in percent of the plot height.
    ///
    /// Without it, one beer in a month of heavy drinking would round to a bar of
    /// zero pixels and look like abstinence.
    public static let minimumVisibleBar = 2.0

    /// Headroom above the tallest bar of the trend chart, so neither it nor the
    /// dashed limit line touches the top edge.
    public static let trendHeadroom = 1.1

    /// Headroom above the tallest bar of the hour and weekday charts, which print
    /// their value above the bar and need room for the text.
    public static let barChartHeadroom = 1.15

    /// The 100 % mark of the trend chart: the taller of the tallest bar and the
    /// limit line, plus headroom. Pinned by `report-chart.json` (`ceiling`);
    /// until v0.86.0 Android wrote `* 1.1` inline in its builder while this side
    /// multiplied by the constants above in three places.
    public static func trendCeiling(tallest: Double, limit: Double) -> Double {
        Swift.max(tallest, limit) * trendHeadroom
    }

    /// The 100 % mark of the hour and weekday charts: the tallest bar plus headroom.
    public static func barCeiling(tallest: Double) -> Double {
        tallest * barChartHeadroom
    }

    /// `value` as a percentage of `max`; zero when `max` is not positive.
    public static func percent(value: Double, max: Double) -> Double {
        max > 0 ? value / max * 100.0 : 0.0
    }

    /// The height of one bar, in percent of the plot.
    ///
    /// `nil` and zero both draw nothing: a weekday that never occurred and a
    /// weekday that was dry are different facts, but they draw the same bar. The
    /// distinction is carried by the value printed above it, not by the bar.
    public static func barHeight(value: Double?, ceiling: Double) -> Double {
        guard let value, value > 0.0 else { return 0.0 }
        return Swift.max(percent(value: value, max: ceiling), minimumVisibleBar)
    }

    /// Labels the report's trend axis carries once a series outgrows twelve bars.
    public static let reportAxisLabels = 8

    /// Labels the Statistics screen's chart carries once a series outgrows twelve bars.
    public static let screenAxisLabels = 6

    /// Which buckets carry an x-axis label.
    ///
    /// Up to twelve buckets, all of them. Beyond that, roughly `target` evenly
    /// spaced ones, always including the first and the last. The report asks for
    /// `reportAxisLabels`, the Statistics screen for `screenAxisLabels`; until
    /// v0.86.0 the screen kept a copy of this loop in its view, in `Double`.
    ///
    /// THE STEP IS A 32-BIT FLOAT, deliberately. Kotlin writes
    /// `((n - 1).toFloat() / (target - 1))`, and the truncation that follows lands
    /// on different indices than the same arithmetic in `Double` would — for 16 of
    /// the first 400 series lengths, `n = 32` among them, which is a month of daily
    /// buckets. Using `Double` here would print a different axis on iOS than on
    /// Android for the same drinking. The shared vectors would catch it.
    public static func labelIndices(count: Int, target: Int = reportAxisLabels) -> [Int] {
        guard count > 0 else { return [] }
        guard count > 12 else { return Array(0..<count) }

        let step = Swift.max(Float(count - 1) / Float(target - 1), 1.0)

        var indices = Set<Int>()
        for slot in 0..<target {
            indices.insert(Swift.min(Int(Float(slot) * step), count - 1))
        }
        indices.insert(count - 1)
        return indices.sorted()
    }

    // ── The donut ────────────────────────────────────────────────────────────

    /// One ring segment, as the three SVG attributes the template expects.
    public struct DonutSlice: Sendable, Equatable {
        /// The dash length, which equals the slice's percentage.
        public let dash: String
        /// The gap that follows it.
        public let gap: String
        /// The rotation that puts this slice after the ones before it.
        public let offset: String
    }

    /// Turns cumulative percentages into `stroke-dasharray` segments.
    ///
    /// The classic trick: a circle of radius 15.9155 has a circumference of very
    /// nearly 100, so a slice's dash length IS its percentage. The offset
    /// `25 - cumulative` rotates the ring to start at twelve o'clock and fill
    /// clockwise.
    ///
    /// Fractions come from grams, not from the rounded integer percents, so the
    /// segments butt up exactly instead of leaving a hairline gap at the end.
    public static func donutSlices(fractions: [Double]) -> [DonutSlice] {
        var cumulative = 0.0
        var slices: [DonutSlice] = []

        for fraction in fractions {
            slices.append(DonutSlice(
                dash: svgNumber(fraction),
                gap: svgNumber(100.0 - fraction),
                offset: svgNumber(25.0 - cumulative)
            ))
            cumulative += fraction
        }
        return slices
    }

    /// Two decimals with a DOT, whatever the device's locale.
    ///
    /// SVG treats both `,` and ` ` as list separators. A locale-aware formatter on
    /// a German device would emit `stroke-dasharray="40,00 60,00"`, which the
    /// renderer reads as the four values `40 0 60 0` — a zero gap, and the ring
    /// paints solid. Kotlin hits this through `String.format()` and passes
    /// `Locale.ROOT` to escape it.
    ///
    /// Swift's `String(format:)` already formats POSIX when given no locale, so the
    /// argument below is redundant. It is passed anyway: the requirement is that
    /// this number never follows the reader's locale, and a requirement stated in
    /// the code outlives one stated in a comment.
    public static func svgNumber(_ value: Double) -> String {
        String(format: "%.2f", locale: Locale(identifier: "en_US_POSIX"), value)
    }
}

// =============================================================================
// ReportPalette
// =============================================================================

public enum ReportPalette {

    /// The donut colour of a stored category name.
    ///
    /// Matches the on-screen palette, so the PDF and the app colour the same
    /// drinking the same way. Hex only, with no character that HTML escaping would
    /// touch, so it can flow through `Template` into an SVG `stroke` attribute.
    public static func color(forCategory name: String) -> String {
        switch name {
        case "BEER": return "#F59E0B"       // amber-500
        case "WINE": return "#9333EA"       // purple-600
        case "SPIRITS": return "#EF4444"    // red-500
        case "LONGDRINK": return "#3B82F6"  // blue-500
        case "LIQUEUR": return "#10B981"    // emerald-500
        default: return "#6B7280"           // gray-500, and OTHER
        }
    }
}

File: android/app/src/main/kotlin/de/godisch/potillus/domain/ReportChart.kt
/* vim: set et ts=4:
 * =============================================================================
 * Libellus Potionis - Privacy-Friendly Alcohol Tracker
 * Copyright (c) 2026 Martin A. Godisch <martin@godisch.de>
 * =============================================================================
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 * In addition, as permitted by section 7 of the GNU General Public License,
 * this program may carry additional permissions; any such permissions that
 * apply to it are stated in the accompanying COPYING.md file.
 *
 * =============================================================================
 */
package de.godisch.potillus.domain

/**
 * The report chart's presentation arithmetic: bar heights, axis labels and the
 * donut's ring geometry.
 *
 * None of this touches Android. It sits here rather than in the PDF builder
 * because it decides what the chart CLAIMS -- a bar of zero height reads as
 * abstinence, a slice that starts a hair late leaves a gap in the ring -- and
 * because both platforms must decide it the same way. The Swift twin is
 * `Domain/ReportChart.swift` in PotillusKit; `test-vectors/report-chart.json`
 * holds them together.
 */
object ReportChart {

    /** Percentage of [value] relative to [max] (0 when [max] is non-positive). */
    fun percent(value: Double, max: Double): Double = if (max > 0) value / max * 100.0 else 0.0

    /**
     * The smallest bar a non-zero value may draw, in percent of the plot height.
     *
     * Without a floor, one beer in a month of heavy drinking scales to a bar of
     * zero pixels and reads as abstinence. Two percent is the smallest strip the
     * print resolution still separates from the baseline.
     */
    const val MINIMUM_VISIBLE_BAR = 2.0

    /**
     * Height of one chart bar, in percent of the plot area.
     *
     * The rule the three bar charts of the report share: a bucket with no value
     * and a bucket with a zero value both draw nothing, and any amount above zero
     * draws at least [MINIMUM_VISIBLE_BAR].
     *
     * NULL AND ZERO DRAW THE SAME BAR, AND THAT IS DELIBERATE. A weekday that
     * never occurred in the period and a weekday that occurred and stayed dry are
     * different facts, but a bar of height zero is the honest picture of both.
     * What tells them apart is the value printed above the bar, which the callers
     * leave blank in the first case.
     *
     * A function rather than three inline expressions: the rule used to be
     * written out once per chart (trend, hour, weekday), which is three chances
     * for it to drift and no way for `ReportChartVectorTest` to reach it. `test-vectors/report-chart.json` holds
     * the `barHeight` cases that the Swift `ReportChart.barHeight` is pinned
     * against, and this is the function that lets the Kotlin suite assert the
     * same ones (0.85.0 QA round).
     *
     * @param value   The bucket's value, or null when the bucket holds no value.
     * @param ceiling The value the full plot height stands for; see the headroom
     *                factors at the call sites.
     */
    fun barHeight(value: Double?, ceiling: Double): Double {
        if (value == null || value <= 0.0) return 0.0
        return percent(value, ceiling).coerceAtLeast(MINIMUM_VISIBLE_BAR)
    }

    /**
     * Two decimals with a DOT, whatever the device's locale.
     *
     * SVG treats both `,` and ` ` as list separators, so a locale-aware format on
     * a German device emits `stroke-dasharray="40,00 60,00"`, which the renderer
     * reads as the four values `40 0 60 0` — a zero gap, and the ring paints
     * solid. [java.util.Locale.ROOT] is what keeps the separator a dot. The Swift
     * twin is `ReportChart.svgNumber`.
     */
    fun svgNumber(value: Double): String = String.format(java.util.Locale.ROOT, "%.2f", value)

    /**
     * One ring segment of the category donut, as the three SVG attributes the
     * template expects.
     */
    data class DonutSlice(val dash: String, val gap: String, val offset: String)

    /**
     * Turns a list of slice percentages into `stroke-dasharray` segments.
     *
     * The classic trick: a circle of radius 15.9155 has a circumference of very
     * nearly 100, so a slice's dash length IS its percentage, and the gap is what
     * remains of the ring. The offset `25 − cumulative` rotates each slice past
     * the ones before it and puts the first one at twelve o'clock, so the ring
     * fills clockwise.
     *
     * The caller passes fractions derived from GRAMS rather than from the rounded
     * integer percents the table prints, so the segments butt up exactly instead
     * of leaving a hairline gap at the end of the ring.
     *
     * The `donut` section of `test-vectors/report-chart.json` pins this geometry
     * on both platforms, and the Kotlin suite needs a named function to assert
     * it against.
     *
     * @param fractions Slice percentages in drawing order, each in 0..100.
     */
    fun donutSlices(fractions: List<Double>): List<DonutSlice> {
        var cumulative = 0.0
        return fractions.map { fraction ->
            val slice = DonutSlice(
                dash = svgNumber(fraction),
                gap = svgNumber(100.0 - fraction),
                offset = svgNumber(25.0 - cumulative),
            )
            cumulative += fraction
            slice
        }
    }

    /**
     * Indices of the buckets that should carry an x-axis label. For a short
     * series (≤ 12 bars) every bucket is labelled; for longer series a small,
     * evenly spaced subset (~8 labels) keeps the axis readable. The first and
     * last buckets are always included.
     */
    fun labelIndices(n: Int): Set<Int> {
        if (n <= 0) return emptySet()
        if (n <= 12) return (0 until n).toSet()
        val target = 8
        val step = ((n - 1).toFloat() / (target - 1)).coerceAtLeast(1f)
        return (0 until target)
            .map { (it * step).toInt().coerceAtMost(n - 1) }
            .toSortedSet()
            .apply { add(n - 1) }
    }
}

/**
 * The donut's category colours.
 *
 * Separate from [ReportChart] because it is a palette rather than arithmetic,
 * the same split the Swift side makes with `ReportPalette`.
 */
object ReportPalette {

    /**
     * Hex colour for a [DrinkCategory] name, matching the on-screen donut palette
     * (`ui.component.categoryColors`) so the PDF donut and the app
     * use the same colours. Escape-safe (no `< > & " '`), so it can flow through
     * SimpleTemplate into an SVG `stroke`/CSS `background`.
     *
     * ReportChartVectorTest checks these colours against the shared vectors the
     * Swift renderer reads too. They must match, or the same drinking prints in
     * different colours on the two platforms.
     */
    fun color(name: String): String = when (name) {
        "BEER" -> "#F59E0B" // amber-500
        "WINE" -> "#9333EA" // purple-600
        "SPIRITS" -> "#EF4444" // red-500
        "LONGDRINK" -> "#3B82F6" // blue-500
        "LIQUEUR" -> "#10B981" // emerald-500
        else -> "#6B7280" // gray-500 (OTHER)
    }
}
